### ISSUE TITLE
GcmRegistrationService starting in background is causing a crash and killing the app

### ISSUE THREAD

**Issue by JonnyBurger** [expo/expo#4207]
## 🐛 Bug Report

ExpoKit for Android includes a Service called `GcmRegistrationService`.
This service is sometimes getting launched from the background, however since Android 8 you need to use a foreground service (which displays a notification to the user) to do that.
Since Expo is trying to start the service from the background, a fatal exception is thrown and the *app gets killed*!

This is pretty subtle, because this never happens while the user has the app open, only when it is in the background. This has probably slipped under the radar because one might assume the app did not crash but did restart because the phone was running low on memory.

I stumbled upon this issue because it would kill my background location service.

```
FATAL EXCEPTION: main
Process: foronered.myapp, PID: 31756
java.lang.RuntimeException: Unable to resume activity {foronered.myapp/host.exp.exponent.MainActivity}: java.lang.IllegalStateException: Not allowed to start service Intent { cmp=foronered.myapp/host.exp.exponent.gcm.GcmRegistrationIntentService }: app is in background uid UidRecord{461a577 u0a267 TRNB bg:+8m16s20ms idle change:uncached procs:1 proclist:31756, seq(1,1,1)}
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4008)
	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4040)
	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.j
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(Trans
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1948)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7045)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)
Caused by: java.lang.IllegalStateException: Not allowed to start service Intent { c
	at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1666)
	at android.app.ContextImpl.startService(ContextImpl.java:1611)
	at android.content.ContextWrapper.startService(ContextWrapper.java:677)
	at host.exp.exponent.experience.BaseExperienceActivity.registerForNotifications(
	at host.exp.exponent.experience.ExperienceActivity.onResume(ExperienceActivity.j
	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1412)
	at android.app.Activity.performResume(Activity.java:7558)
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4000)
```

### Environment

ExpoKit v32
Android 9
Samsung Galaxy S8

### Steps to Reproduce

Not sure. I am not using Google Cloud Messaging at all, but have logged the exception through `adb logcat`

**Comment by zanechua** [comment]
Can confirm that i am seeing this in the latest expokit v33.

**Comment by andersravn** [comment]
We have a crash rate of 1.32% the last 30 days. The vast majority of those produce the same stack trace as above. They all seem to be happening on Android 9. We are using ExpoKit SDK 30.

**Comment by jamesbechet** [comment]
I'm seeing this same stack trace with SDK 32 non-detached.

**Comment by briefjudofox** [comment]
I also noticed this exception when troubleshooting why our geofence task never gets called
 when the app is closed (works fine when foregrounded and backgrounded).  Based on the [android-play-location sample](https://github.com/googlesamples/android-play-location/blob/master/Geofencing/app/src/main/AndroidManifest.xml) I've had some luck getting our app to respond to geofence events while in a closed state by setting `export` to `true` on the `TaskBroadcastReceiver` in the expo-task-manager module [manifest](https://github.com/expo/expo/blob/master/packages/expo-task-manager/android/src/main/AndroidManifest.xml).

After ejecting and changing the expo-task-manager module manifest the compiled app's manifest looks like this:

 ```xml
...
<receiver android:enabled="true" android:exported="true" android:name="expo.modules.taskManager.TaskBroadcastReceiver">
    <intent-filter>
        <action android:name="expo.modules.taskManager.TaskBroadcastReceiver.INTENT_ACTION"/>
        <action android:name="android.intent.action.BOOT_COMPLETED"/>
    </intent-filter>
</receiver>
...
```
instead of this:
```xml
...
<receiver android:exported="false" android:name="expo.modules.taskManager.TaskBroadcastReceiver">
    <intent-filter>
        <action android:name="expo.modules.taskManager.TaskBroadcastReceiver.INTENT_ACTION"/>
        <action android:name="android.intent.action.BOOT_COMPLETED"/>
    </intent-filter>
</receiver>
...
```

First, I'm curious if changing this fixes anything for you all.  If so, perhaps some variant of this change can be made on the module itself?

**Comment by jamesbechet** [comment]
I'm still seeing this error with SDK 34

**Comment by cem2ran** [comment]
@tsapeta I'd be happy to provide more info. We're seeing `illegalStateException` when the service is started. Perhaps related?

If you think it's a separate issue I'd be happy to open a new issue.

We are seeing two seperate crash clusters in Google Play Android vitals, posted below is the largest:

Stacktrace:
```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3993)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:4025)
  at android.app.servertransaction.ResumeActivityItem.execute (ResumeActivityItem.java:51)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:145)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:70)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1926)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:6986)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1445)
Caused by: java.lang.IllegalStateException: 
  at android.app.ContextImpl.startServiceCommon (ContextImpl.java:1666)
  at android.app.ContextImpl.startService (ContextImpl.java:1611)
  at android.content.ContextWrapper.startService (ContextWrapper.java:677)
  at host.exp.exponent.experience.BaseExperienceActivity.registerForNotifications (BaseExperienceActivity.java:258)
  at host.exp.exponent.experience.ExperienceActivity.onResume (ExperienceActivity.java:256)
  at android.app.Instrumentation.callActivityOnResume (Instrumentation.java:1412)
  at android.app.Activity.performResume (Activity.java:7571)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3985)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:4025)
  at android.app.servertransaction.ResumeActivityItem.execute (ResumeActivityItem.java:51)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:145)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:70)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1926)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:6986)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1445)
```

Make | Devices | Percentage
--- | --- | ---
Galaxy S9 (starlte) | 21 | 17.5%
Galaxy S8 (dreamlte) | 14 | 11.7%
Galaxy S10+ (beyond2) | 10 | 8.3%
Motorola one (deen_sprout) | 9 | 7.5%
Galaxy S10 (beyond1) | 9 | 7.5%
Galaxy S9+ (star2lte) | 8 | 6.7%
OnePlus 6 (OnePlus6) | 8 | 6.7%
Galaxy Note9 (crownlte) | 7 | 5.8%
OnePlus6T (OnePlus6T) | 5 | 4.2%
Galaxy A8(2018) (jackpotlte) | 4 | 3.3%
OnePlus 7 (OnePlus7) | 4 | 3.3%
Galaxy A7 (2018) (a7y18lte) | 4 | 3.3%
Pixel 2 XL (taimen) | 2 | 1.7%
Galaxy A50 (a50) | 2 | 1.7%
greatlte | 2 | 1.7%
Galaxy S8+ (dream2lte) | 1 | 0.8%
Galaxy A40 (a40) | 1 | 0.8%
Galaxy A20e (a20e) | 1 | 0.8%
OnePlus3 (OnePlus3) | 1 | 0.8%
OnePlus 7 Pro (OnePlus7Pro) | 1 | 0.8%
Others | 6 | 5.0%

**Comment by thomasgrastein** [comment]
Seeing the exact same error on Android 9, most frequently on the Galaxy S9 aswell. 
Our app is only using location with MapView from react-native-maps while the app is open.

**Comment by jacobarvidsson** [comment]
I am also seeing this issue.

**Comment by alex-fournier** [comment]
Same error as @cem2ran recently (Galaxy s8 / s9 on Android 9).
I'm on Expo SDK 33.

**Comment by caioedut** [comment]
Same issue here.

**Comment by huguesn** [comment]
This bug occurs only when **FCM** (Firebase Cloud Messaging) is not enabled.
https://github.com/expo/expo/blob/882f3e5ef9be1e8da4abd28e2952ee01b8e3233e/android/expoview/src/main/java/host/exp/exponent/experience/BaseExperienceActivity.java#L256-L259

Google deprecated **GCM** (Google Cloud Messaging) in favor of **FCM** on April, and the Expo team is still not enabling FCM by default because it requires extra steps like setting up a Firebase account ([they made a post about it](https://blog.expo.io/android-gcm-notifications-are-going-away-7fc2ef5fa3e9))

So.... until it get fixed, we have no choice but to enable FCM if we want this problem to go away, and this is the future anyway as GCM is deprecated.

Grabbing a ☕️, let's do this! https://docs.expo.io/versions/v35.0.0/guides/using-fcm/

**Comment by andersravn** [comment]
I guess I'll try and remove it from our ExpoKit project then, as we don't use it.

**Comment by pyankoff** [comment]
Anyone managed to fix the issue? 

Removing GCM and enabling FCM doesn't seem to help.

**Comment by huguesn** [comment]
Good news, it works for me! 

I redeployed my app **with FCM enabled** 5 days ago (https://github.com/expo/expo/issues/4207#issuecomment-534081721) and I've not seen a single crash yet related to GcmRegistrationService. Even more, my app launched 2 days ago with +1k download, and still nothing.

**Comment by Gorbus** [comment]
Is there an easiest way to solve it using the managed workflow when we don't need any notifications? 
I am not using Firebase or anything so it seems a bit overkill to configure this to remove this crash but the crash is happening here as well.

**Comment by pyankoff** [comment]
Maybe the crash is not related to GCM?
Seems like it's happening on both managed and ejected builds: https://forums.expo.io/t/android-production-build-high-crash-rate/28444

**Comment by esamelson** [comment]
Hey all -- apologies for the silence here. We'll hopefully have a fix out for this in the next day or so. In the meantime, if you can't wait for a fix, enabling FCM and rebuilding will fix the crash. You'll need to make a new build with `expo build:android` for this to take effect.

**Comment by daniel** [comment]
I get this error even though I'm not using GCM or push notifications in any way. Do I still need to enable FCM?

**Comment by pyankoff** [comment]
Mine was an issue not related to GCM. JS errors now crash the app in SDK35 - https://forums.expo.io/t/high-crash-rate-on-android-sdk35/29285/2?u=andrey

**Comment by esamelson** [comment]
We've deployed a fix for this to production for SDK 35, so if you rebuild your app with `expo build:android`, install `expokit@35.0.3`, or build your app with `turtle-cli@0.12.7` or later this particular crash should go away. This only applies to the crash that is the topic of this issue, the IllegalStateException which was caused by legacy GCM code. There is no need anymore to enable FCM in order to fix this crash, you just need to rebuild your app.
